cpp/src/gandiva/engine.cc

```diff
@@ -152,6 +153,7 @@ Status Engine::LoadFunctionIRs() {
   if (!functions_loaded_) {
     ARROW_RETURN_NOT_OK(LoadPreCompiledIR());
     ARROW_RETURN_NOT_OK(DecimalIR::AddFunctions(this));
+    ARROW_RETURN_NOT_OK(LoadExternalPreCompiledIR());
```

Comment:
Can we add `Engine::AddPreCompiledID()` instead of adding `LoadExternalPreCompiledIR()` (and `LLVMIRStore`) and calling it here?

Reply:
Could you please explain with more details?

Do you mean we add a new API called `AddPreCompiledIR` (do you mean `IR` instead of `ID`?) to the `Engine` class so that developers can call this API directly to add a third party IR? 

Currently, the `Engine` class is not used by Gandiva users directly, typically Gandiva users uses the `Projector` API, which internally calls the `LLVMGenerator` which internally calls the `Engine` API. For example:
```
std::shared_ptr<Projector> projector;
  ARROW_EXPECT_OK(Projector::Make(schema, {multiply}, TestConfiguration(), &projector));
...
ARROW_EXPECT_OK(projector->Evaluate(*in_batch, pool_, &outs));
```
So if we add `AddPreCompiledIR` to the `Engine` class, this will be a new API surface for users as well.

In this PR, it is expected Gandiva users to use the new `LLVMIRStore` API to load third party IR into Gandiva, it is possible we move this API to `Engine` class, but since the `Engine` class is not a global singleton, and is re-instantiated every time by the `LLVMGenerator`, if we store IR in `Engine`, we may have to introduce similar global states in `Engine` like we do for `LLVMIRStore`. 

Please let me know if I understand correctly and advise which approach we prefer. Thanks.

Reply:
> Could you please explain with more details?

Sure! Sorry for missing information...
 
> Do you mean we add a new API called `AddPreCompiledIR` (do you mean `IR` instead of `ID`?) to the `Engine` class so that developers can call this API directly to add a third party IR?

Yes. (It's a typo. Sorry...)

> Currently, the `Engine` class is not used by Gandiva users directly, typically Gandiva users uses the `Projector` API, which internally calls the `LLVMGenerator` which internally calls the `Engine` API. For example:
> 
> ```
> std::shared_ptr<Projector> projector;
>   ARROW_EXPECT_OK(Projector::Make(schema, {multiply}, TestConfiguration(), &projector));
> ...
> ARROW_EXPECT_OK(projector->Evaluate(*in_batch, pool_, &outs));
> ```
> 
> So if we add `AddPreCompiledIR` to the `Engine` class, this will be a new API surface for users as well.
> 
> In this PR, it is expected Gandiva users to use the new `LLVMIRStore` API to load third party IR into Gandiva, it is possible we move this API to `Engine` class, but since the `Engine` class is not a global singleton, and is re-instantiated every time by the `LLVMGenerator`, if we store IR in `Engine`, we may have to introduce similar global states in `Engine` like we do for `LLVMIRStore`.

How about adding `FunctionRegistry::Add(NativeFunction func, const std::string& bitcode_path)` and using it and `Engine::AddPreCompiledIR()` in `LLVMGenerator::Add()`?

Then, users can only use `FunctionRegistry::Add()` to add an external function instead of using `FunctionRegistry::AddFunction()` and LVMIRStore`.

Reply:
> How about adding FunctionRegistry::Add(NativeFunction func, const std::string& bitcode_path) and using it and Engine::AddPreCompiledIR() in LLVMGenerator::Add()?

That seems reasonable, users typically have to do both (registering the function meta and providing function bitcode) for such purpose, so it is easier to make them call single API to achieve it.
But there are two issues we may consider:
1) One issue is `FunctionRegistry::Add(NativeFunction func, const std::string& bitcode_path)` this API makes function metadata to be 1-1 mapped to the bitcode, however, for a given bitcode file, it may contain bitcodes for multiple functions. But it provides additional information where a function's bitcode can be found if we have such an API, which may help to avoid linking unnecessary modules if a function is never used in the expression (see `Performance` section in this issue description)
2) the other issue is this PR only provides the capability to register a pre-compiled function via LLVM IR/bitcode, but I think it is possible to use the `AddFunction` API and some shared library mechanism to support non pre-compiled functions as well, so it may not be mandatory to register a function metadata with bitcode (in the future)

I will take a further look to see how this could be improved. At the same time, @js8544 do you have any feedback on the APIs here? Thanks.

Reply:
I agree with you that splitting function declaration and implementation is better for future maintenance. I have some  other comments below.

Reply:
Ah, I see. Thanks for organizing these issues.

1. I like the `std::vector<NativeFunction>` approach. I think that managing both of signatures and bitcodes by `FunctionRegistry` is better than introducing IR store for bitcodes. The signature and bitcode mapping will be useful for loading bitcode to `gandiva::Engine` selectively for performance.
2. Could you explain "some shared library mechanism" a bit more? I think that we need to load functions as bitcode to `gandiva::Engine` to use them. Does "some shared library mechanism" mean that we may be able to add support for loading native functions (sorry, I don't know how to call it. I imagine that `gcc -shared -o multiply_by_two.so multiply_by_two.cc` and `handle = dlopen("multiply_by_two.so"); dlsym(handle, "multiply_by_two_int32")`) to `gandiva::Engine`?

Reply:
> I like the std::vector<NativeFunction> approach. I think that managing both of signatures and bitcodes by FunctionRegistry is better than introducing IR store for bitcodes. The signature and bitcode mapping will be useful for loading bitcode to gandiva::Engine selectively for performance

I agree. It seems simpler for users to call just one API because I don't see any use scenario that users would only call one of the two APIs (`FunctionRegistry::Add`, `LLVMExternalIRStore::Add`), additionally, we could take advantage of this signature <==> bitcode mapping later to optimize performance during LLVM module construction. @js8544 do you have any opinion on this idea? Thanks.

> Could you explain "some shared library mechanism" a bit more?
> I imagine that gcc -shared -o multiply_by_two.so multiply_by_two.cc and handle = dlopen("multiply_by_two.so"); dlsym(handle, "multiply_by_two_int32")) to gandiva::Engine

Sorry I don't make it clear enough. This probably worths another PR for it and I can add some background here. This PR allows users to call pre-compiled functions from LLVM IR/bitcode. Gandiva users need to generate LLVM IR for their functions for integration. This approach will work for small and simple functions, typically for functions without using any dependent library. However, if a function would like to do some more complex task, typically function authors would like to rely on third party libraries. For example, if I would like to implement a function calling a service via HTTP with JSON request/response (e.g. twitter API/OpenAI API), this function will have to depend on an HTTP library and a JSON parsing library (or even some async IO library). If all these dependent libraries are compiled and generated into LLVM IR, the IR/bitcode will be non trivial, and this will make LLVM module construction pretty slow. 

So I think it may help to compile this kind of complex function into a shared library (instead of LLVM IR), and ask the LLVM engine to load the shared library during runtime and lookup function directly in the shared library. I briefly look into how this could be done, but more investigation is needed later:
1) LLVM has an API called `DynamicLibrary::LoadLibraryPermanently` which can be used for loading a shared library during runtime. 
1.1) this is the `DynamicLibrary` API doc in LLVM, https://llvm.org/doxygen/classllvm_1_1sys_1_1DynamicLibrary.html
1.2) `Engine::InitOnce` function in `engine.cc:115` currently calls this `LoadLibraryPermanently` API but it loads nothing currently.
2) As you said, we can use `dlopen` and `dlsym` to get function pointer in the shared library
3) `Engine::AddGlobalMappingForFunc` is used for programmatically setting up a mapping from function name to function pointer, currently, there are some functions like `gdv_fn_context_arena_malloc` is mapped via this API
4) `void* Engine::CompiledFunction(llvm::Function* irFunction)` is used for looking up the function in the module
I think it is possible to piece all these together to make them work and address the above complex function performance issue.